Only process relevant transactions

grod220 opened the pull request:
Closes https://github.com/penumbra-zone/web/issues/1792

We should only be processing transactions that have nullifiers, spends, or swaps we recognize. At the moment, if there is a relevant tx in a block, we later process the entire block. This caused the original issue of an LP position misattributed to the owner when it was external.

Note, we need to version the DB given users may have in their db's LP NFTs and Auction NFTs that happened to be in the same block as another person's.

TalDerei reviewed (approved):
nice catch. Is bumping the idb version necessary, since we haven't modified the database schema?

grod220 commented:
=== Part 2 ===

The original simple bug fix actually exposed another bug [here](https://github.com/penumbra-zone/web/pull/1793/files#diff-ccb0493c986b958cf859e773170457e5ff59a4a2e0d742698492247f6ae872daL488). Adding to that map duplicate entries caused multiple attempts to save the same auction state (certain invalid combos raise exceptions).

Extracted `identifyTransactions()` to a helper file, refactored it, patched the bug, and added tests.

grod220 commented on packages/query/src/block-processor.ts:
```diff
@@ -129,9 +118,7 @@ export class BlockProcessor implements BlockProcessorInterface {
-        if (globalThis.__DEV__) {
-          console.debug('Sync failure', attemptNumber, e);
-        }
+        console.error(`Sync failure #${attemptNumber}: `, e);
```
Think we shouldn't require a special flag to see sync failures. Will help with debugging if users get stuck.

grod220 commented on packages/query/src/block-processor.ts:
```diff
@@ -450,64 +432,6 @@ export class BlockProcessor implements BlockProcessorInterface {
-
-      for (const spentNullifier of spentNullifiers) {
-        if (txNullifiers.some(txNullifier => spentNullifier.equals(txNullifier))) {
-          txId = new TransactionId({ inner: await sha256Hash(tx.toBinary()) });
```
This line was the source of the bug. Because the key of the map is an object, when you create a new object (even though it has the same values), it will be added as a new entry in the map. Later on, the redundant entries in the map can cause problems. So this should have been like the loop below:

```diff
- txId = new TransactionId({ inner: await sha256Hash(tx.toBinary()) });
+ txId ??= new TransactionId({ inner: await sha256Hash(tx.toBinary()) });
```

grod220 commented on packages/query/src/block-processor.ts:
```diff
@@ -450,64 +432,6 @@ export class BlockProcessor implements BlockProcessorInterface {
-        if (txNullifiers.some(txNullifier => spentNullifier.equals(txNullifier))) {
-          txId = new TransactionId({ inner: await sha256Hash(tx.toBinary()) });
-          relevantTx.set(txId, tx);
-          spentNullifiers.delete(spentNullifier);
```
We seem to be mutating the inputs `spentNullifiers` and `commitmentRecordsByStateCommitment`. This seems like a brittle pattern, especially from within a loop that iterates over that object. I could not see a reason why we needed to do this.

grod220 commented on packages/query/src/helpers/identify-txs.ts:
```diff
@@ -0,0 +1,137 @@
+  const relevantTxs: RelevantTx[] = [];
+  const recoveredSourceRecords: RecoveredSourceRecords = [];
+
+  const searchPromises = blockTx.map(tx => searchRelevant(tx, spentNullifiers, commitmentRecords));
+  const results = await Promise.all(searchPromises);
```
New feature: we now process transactions in parallel versus the previous synchronous loop

TalDerei replied:
Is there a sense of how the concurrency performs when handling large blocks with many transactions? Is the reduction in unnecessary computations and storage operations pronounced when focusing on processing only relevant transactions?

grod220 replied:
> handling large blocks with many transactions

This should perform better as it's no longer sequentially peforming `await sha256Hash(tx.toBinary())`, but it's happening in parallel. However, the hashing cost may be trivial and this optimization may only be detectable with a very large amount of txs. 

> impact of avoiding unnecessary computations and storage operations

This code impacts this by making the storage of irrelevant transactions unnecessary. In a situation with quite large blocks with a ton of transactions, this would likely be noticeable at some point had we not fixed it.

grod220 commented on packages/storage/src/indexed-db/config.ts:
```diff
@@ -2,4 +2,4 @@
  * database upgrades and ensure that the correct schema version is applied.
  */
-export const IDB_VERSION = 46;
+export const IDB_VERSION = 47;
```
Necessary to bump storage version as users may have NFTs (auction/LP positions) that are misattributed to them as being the owner.

TalDerei commented on packages/query/src/block-processor.ts:
```diff
@@ -318,28 +305,23 @@ export class BlockProcessor implements BlockProcessorInterface {
       // this is a network query
       const blockTx = await this.querier.app.txsByHeight(compactBlock.height);
 
-      // identify tx that involve a new record
-      // - compare nullifiers
-      // - compare state commitments
-      // - collect relevant tx for info generation later
-      // - if matched by commitment, collect record with recovered source
```
reason for removing code comment?

grod220 replied:
Carried over to external function. But will add a code comment above where its used again.

TalDerei commented on packages/query/src/helpers/identify-txs.ts:
```diff
@@ -0,0 +1,137 @@
+// also returns records with recovered sources
+export const identifyTransactions = async (
+  spentNullifiers: Set<Nullifier>,
+  commitmentRecords: Map<StateCommitment, SpendableNoteRecord | SwapRecord>,
```
think we might have another potential bug. I don’t think we've ever addressed the case where the map holding `StateCommitment` objects contains [duplicate commitments](https://github.com/penumbra-zone/penumbra/issues/3945)?

grod220 replied:
In that case, the worst that could happen is in this line: https://github.com/penumbra-zone/web/pull/1793/files#diff-df9a3308a5ae26e8f5e912c4bdd659a76d12e6f751ec7c3eb8de3c0b81147bf0R91-R98 where there are two duplicate recovered records. And in that case, we just [save to the database](https://github.com/penumbra-zone/web/blob/75ef4f5bb7bb07c4ea0478bee360beba12220f28/packages/query/src/block-processor.ts#L424-L440) twice. Don't see a failure case we need to worry about.
